database/manager.go

```diff
@@ -0,0 +1,181 @@
+
+// Run writes all committed keys to the database on the configured timer.
+// This function should only be called once.
+func (dm *DatabaseManager) Run() {
+	for range time.Tick(dm.interval) {
+		for id, km := range dm.keyManagers {
+			// Ensure we're not writing the default value
+			if km.maxCommittedHeight > 0 {
+				storedHeight, err := getLatestProcessedBlockHeight(dm.db, id)
+				if err != nil && !IsKeyNotFoundError(err) {
+					dm.logger.Error("Failed to get latest processed block height", zap.Error(err))
+					continue
+				}
+				if storedHeight >= km.maxCommittedHeight {
+					continue
+				}
+				dm.logger.Debug(
+					"writing height",
+					zap.Uint64("height", km.maxCommittedHeight),
+					zap.String("relayerID", id.ID.String()),
+				)
+				err = dm.db.Put(id.ID, LatestProcessedBlockKey, []byte(strconv.FormatUint(km.maxCommittedHeight, 10)))
+				if err != nil {
+					dm.logger.Error("Failed to write latest processed block height", zap.Error(err))
+					continue
+				}
+			}
+		}
+	}
+}
+
+func (dm *DatabaseManager) Get(id RelayerID, key DataKey) ([]byte, error) {
+	return dm.db.Get(id.ID, key)
+}
+
+func (dm *DatabaseManager) PrepareHeight(id RelayerID, height uint64, totalMessages uint64) error {
+	km, ok := dm.keyManagers[id]
+	if !ok {
+		dm.logger.Error("Key manager not found", zap.String("relayerID", id.ID.String()))
+		return fmt.Errorf("key manager not found")
+	}
+	km.prepareHeight(height, totalMessages)
+	return nil
+}
+
+func (dm *DatabaseManager) Finished(id RelayerID, height uint64) {
+	km, ok := dm.keyManagers[id]
+	if !ok {
+		// This is not an error, as this will occur if the database is not eligible for writing
+		dm.logger.Debug("Key manager not found", zap.String("relayerID", id.ID.String()))
+		return
+	}
+	km.finished <- height
+}
+
+//
+// keyManager commits keys to be written to the database in a thread safe manner.
+//
+
+type keyManager struct {
+	logger                   logging.Logger
+	queuedHeightsLock        *sync.Mutex
+	queuedHeightsAndMessages map[uint64]*messageCounter
+	maxHeightLock            *sync.Mutex
+	maxCommittedHeight       uint64
+	finished                 chan uint64
+}
+
+func newKeyManager(logger logging.Logger) *keyManager {
+	return &keyManager{
+		logger:                   logger,
+		queuedHeightsLock:        &sync.Mutex{},
+		queuedHeightsAndMessages: make(map[uint64]*messageCounter),
+		maxHeightLock:            &sync.Mutex{},
+		finished:                 make(chan uint64),
+	}
+}
+
+// Run listens for finished signals from application relayers, and commits the
+// height once all messages have been processed.
+// This function should only be called once.
+func (km *keyManager) run() {
+	for height := range km.finished {
+		counter, ok := km.queuedHeightsAndMessages[height]
+		if !ok {
+			continue
+		}
+		counter.processedMessages++
+		if counter.processedMessages == counter.totalMessages {
+			km.commitHeight(height)
+			delete(km.queuedHeightsAndMessages, height)
+		}
+	}
```

Comment:
So I'm pretty sure we have a race condition in here. Lets say the current highest processed height is 99 and we are currently processing blocks 100 and 101. If 101 finished processing before block 100, we will update the highest processed block to 101, while not having processed 100 yet. We have to check `queuedHeightsAndMessages` to make sure there is no lower height being processed. 

We could do something like... Let's say our highest processed block is N. If the height we're committing is N+1, set the value to that. Otherwise, throw that block height in a min heap. Then when we finally process block N+1, we look in the min heap for N+2, and keep removing values from the heap as long as we have the next block in sequence.

Reply:
If I'm understanding the code correctly, it does look like the commitHeight function needs to do some buffering to avoid gaps in the committed block height.

A circular buffer could be used in place of a min heap if the difference between the highest processed block block and the current height is bounded.

Reply:
That's a great catch. I've implemented the min-heap based approach you suggested.

On a related note, I'm debating how we should handle the case in which a block height never calls `commitHeight`. In that case, the min-heap would grow unbounded, and no new data would be written to the DB. One approach would be to kill the entire application if the min-heap grows too large, and attempt to re-process the missing block on restart. If the issue persists, the operator would have to intervene by either setting `process-missed-blocks` to false, or manually updating the DB. One thing I like about this approach is that all blocks up to the height stored in the DB are guaranteed to be processed; only with manual intervention can this be made untrue.

Reply:
Hmmm... Yeah I wonder what the best approach would be here to handle the case of a missing block. I don't really want to kill the whole application for one missing block for one application.

The approach that is coming to mind is setting some threshold, say 100 blocks, that if the min heap grows bigger than, we just pop the min value from the heap and set that as our most recently processed block, and throw an error for any missed blocks. We could also write the missed blocks to a table somewhere for potential reprocessing later.